src/dlog_with_el_gamal_commitment.rs

```diff
@@ -0,0 +1,509 @@
+        aux: &Aux,
+        data: Data<C>,
+        pdata: PrivateData,
+        security: &SecurityParams,
+        rng: &mut impl rand_core::RngCore,
+    ) -> Result<(Commitment<C>, Proof), Error> {
+        let (comm, pcomm) = super::interactive::commit(data, rng)?;
+        let challenge = challenge::<C, D>(shared_state, aux, data, &comm, security);
+        let proof = super::interactive::prove::<C>(pdata, &pcomm, &challenge)?;
+        Ok((comm, proof))
+    }
+
+    /// Verify the proof, deriving challenge independently from same data
+    pub fn verify<C: Curve, D: Digest>(
+        shared_state: &impl udigest::Digestable,
+        aux: &Aux,
+        data: Data<C>,
+        commitment: &Commitment<C>,
+        security: &SecurityParams,
+        proof: &Proof,
+    ) -> Result<(), InvalidProof> {
+        let challenge = challenge::<C, D>(shared_state, aux, data, commitment, security);
+        super::interactive::verify::<C>(data, commitment, &challenge, proof)
+    }
+
+    /// Deterministically compute challenge based on prior known values in protocol
+    pub fn challenge<C: Curve, D: Digest>(
+        shared_state: &impl udigest::Digestable,
+        aux: &Aux,
+        data: Data<C>,
+        commitment: &Commitment<C>,
+        security: &SecurityParams,
+    ) -> Challenge {
+        let tag = "paillier_zk.dlog_with_el_gamal.ni_challenge";
+        let aux = aux.digest_public_data();
+        let seed = udigest::inline_struct!(tag {
+            shared_state,
+            aux,
+            security,
+            data,
+            commitment,
+        });
+        let mut rng = rand_hash::HashRng::<D, _>::from_seed(seed);
+        super::interactive::challenge(security, &mut rng)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use generic_ec::{Curve, Point, Scalar};
+    use rug::Integer;
+    use sha2::Digest;
+
+    use crate::common::{IntegerExt, InvalidProofReason};
+
+    fn run<R: rand_core::RngCore + rand_core::CryptoRng, C: Curve, D: Digest>(
+        rng: &mut R,
+        security: super::SecurityParams,
+        data: super::Data<C>,
+        pdata: super::PrivateData,
+    ) -> Result<(), crate::common::InvalidProof> {
+        let aux = crate::common::test::aux(rng);
+
+        let shared_state = "shared state";
+
+        let (commitment, proof) =
+            super::non_interactive::prove::<C, D>(&shared_state, &aux, data, pdata, &security, rng)
+                .unwrap();
+        super::non_interactive::verify::<C, D>(
+            &shared_state,
+            &aux,
+            data,
+            &commitment,
+            &security,
+            &proof,
+        )
+    }
+
+    fn passing_test<C: Curve, D: Digest>() {
+        let mut rng = rand_dev::DevRng::new();
+        let security = super::SecurityParams {
+            q: (Integer::ONE << 128_u32).into(),
+        };
+        let y =         Integer::from_rng_pm(&security.q,&mut rng);
+        let lambda = Integer::from_rng_pm(&security.q,&mut rng);
```

Comment:
```suggestion
        let lambda = Integer::from_rng_pm(&Integer::curve_order::<C>(), &mut rng);
```

Since the order is a prime, we can use the `gen_invertable` I belive